File: src/NTIX.CLI/Commands/Commands.cs
using CliFx;
using CliFx.Binding;
using CliFx.Infrastructure;
using NTIX.Core.Config;
using NTIX.Core.StateManagement;
using NTIX.Core.Diff;
using NTIX.Core.Execution;
using NTIX.Core.Lock;
using NTIX.Core;
using NTIX.Core.Models;
using Spectre.Console;
using System.Runtime.Versioning;

namespace NTIX.CLI.Commands;

[Command("apply", Description = "Apply desired state (install/remove packages)")]
public partial class ApplyCommand : ICommand
{
    [CommandParameter(0, Name = "config-path", Description = "Path to configuration file (default: ~/ntix/config.lua)")]
    public string? ConfigPath { get; set; }

    [CommandOption("dry-run", 'd', Description = "Show what would change without applying")]
    public bool DryRun { get; set; }

    [CommandOption("no-gc", Description = "Don't remove packages not in config")]
    public bool NoGc { get; set; }

    [CommandOption("stop-on-failure", Description = "Stop on first package failure instead of continuing")]
    public bool StopOnFailure { get; set; }

    [CommandOption("adopt", 'a', Description = "Adopt already-installed packages into NTIX state")]
    public bool Adopt { get; set; }

    [CommandOption("upgrade", 'u', Description = "Check for and apply available upgrades")]
    public bool Upgrade { get; set; }

    [SupportedOSPlatform("windows")]
    public async ValueTask ExecuteAsync(IConsole console)
    {
        if (!ProcessHelper.IsRunningAsAdmin())
        {
            AnsiConsole.MarkupLine("[red]Error: ntix apply requires administrator privileges.[/]");
            AnsiConsole.MarkupLine("Please re-run in an elevated terminal (Run as Administrator).");
            Environment.ExitCode = 1;
            return;
        }

        var isNew = ConfigPath is null && !File.Exists(ConfigLoader.DefaultConfigPath);
        ConfigPath = ConfigLoader.EnsureDefaultConfig(ConfigPath);
        if (isNew)
        {
            AnsiConsole.MarkupLine($"[green]Created default config at {ConfigPath}[/]");
            AnsiConsole.MarkupLine("Edit it to add your packages, then run [bold]ntix diff[/] again.");
            return;
        }
        var config = ConfigLoader.Load(ConfigPath);
        var state = StateService.LoadState() ?? new NTIX.Core.Models.State();

        DiffResult diff = null!;
        var configFileName = Path.GetFileName(ConfigPath);
        await AnsiConsole.Status()
            .Spinner(Spinner.Known.Dots)
            .SpinnerStyle(Style.Parse("yellow"))
            .StartAsync($"[bold]{configFileName}[/]", async ctx =>
            {
                var progress = new Progress<string>(s => ctx.Status($"[dim]{s}[/]"));
                diff = await DiffEngine.ComputeDiffAsync(config, state, progress: progress, adoptMode: Adopt, upgradeMode: Upgrade);
            });

        var tree = CommandsHelper.BuildDiffTree(configFileName, config, diff);
        AnsiConsole.Write(tree);

        foreach (var w in diff.Warnings)
            AnsiConsole.MarkupLine($"[yellow]Warning: {Markup.Escape(w)}[/]");

        if (NoGc)
            diff.ToRemove.Clear();

        if (DryRun)
        {
            AnsiConsole.MarkupLine("\n[yellow](Dry run - no changes made)[/]");
            return;
        }

        if (diff.IsEmpty)
            return;

        using var lockFile = new LockFile();
        var statePath = StateService.GetStatePath();
        var success = await ExecutionEngine.ApplyDiffAsync(
            diff, config.Options, state, statePath,
            stopOnFailure: StopOnFailure,
            onOutput: Console.WriteLine,
            onError: msg => AnsiConsole.MarkupLine($"[red]{Markup.Escape(msg)}[/]"));

        if (success)
        {
            AnsiConsole.MarkupLine("\n[green]Done.[/]");
        }
        else
        {
            AnsiConsole.MarkupLine("\n[red]Some operations failed.[/]");
            Environment.ExitCode = 1;
        }
    }
}

[Command("diff", Description = "Show what would change")]
public partial class DiffCommand : ICommand
{
    [CommandParameter(0, Name = "config-path", Description = "Path to configuration file (default: ~/ntix/config.lua)")]
    public string? ConfigPath { get; set; }

    [CommandOption("adopt", 'a', Description = "Show packages that would be adopted")]
    public bool Adopt { get; set; }

    [CommandOption("upgrade", 'u', Description = "Check for and apply available upgrades")]
    public bool Upgrade { get; set; }

    public async ValueTask ExecuteAsync(IConsole console)
    {
        var isNew = ConfigPath is null && !File.Exists(ConfigLoader.DefaultConfigPath);
        ConfigPath = ConfigLoader.EnsureDefaultConfig(ConfigPath);
        if (isNew)
        {
            AnsiConsole.MarkupLine($"[green]Created default config at {ConfigPath}[/]");
            AnsiConsole.MarkupLine("Edit it to add your packages, then run [bold]ntix diff[/] again.");
            return;
        }
        var config = ConfigLoader.Load(ConfigPath);
        var state = StateService.LoadState() ?? new NTIX.Core.Models.State();

        DiffResult diff = null!;
        var configFileName = Path.GetFileName(ConfigPath);
        await AnsiConsole.Status()
            .Spinner(Spinner.Known.Dots)
            .SpinnerStyle(Style.Parse("yellow"))
            .StartAsync($"[bold]{configFileName}[/]", async ctx =>
            {
                var progress = new Progress<string>(s => ctx.Status($"[dim]{s}[/]"));
                diff = await DiffEngine.ComputeDiffAsync(config, state, progress: progress, adoptMode: Adopt, upgradeMode: Upgrade);
            });

        var tree = CommandsHelper.BuildDiffTree(configFileName, config, diff);
        AnsiConsole.Write(tree);

        foreach (var w in diff.Warnings)
            AnsiConsole.MarkupLine($"[yellow]Warning: {Markup.Escape(w)}[/]");
    }
}

[Command("state", Description = "Show current NTIX state")]
public partial class StateCommand : ICommand
{
    public async ValueTask ExecuteAsync(IConsole console)
    {
        var state = StateService.LoadState();

        if (state == null)
        {
            AnsiConsole.MarkupLine("[yellow]No state file found.[/]");
            return;
        }

        AnsiConsole.MarkupLine("[bold]NTIX State:[/]");

        if (state.Winget.Count == 0 && state.Chocolatey.Count == 0 && state.Scoop.Count == 0)
        {
            AnsiConsole.MarkupLine("  [dim](empty)[/]");
        }
        else
        {
            foreach (var (id, ver) in state.Winget)
                AnsiConsole.MarkupLine($"  [cyan]winget: {id} ({ver})[/]");

            foreach (var (id, ver) in state.Chocolatey)
                AnsiConsole.MarkupLine($"  [magenta]chocolatey: {id} ({ver})[/]");

            foreach (var (id, ver) in state.Scoop)
                AnsiConsole.MarkupLine($"  [blue]scoop: {id} ({ver})[/]");
        }
    }
}

internal static class CommandsHelper
{
    private static readonly Dictionary<string, string> SourceMarkup = new(StringComparer.OrdinalIgnoreCase)
    {
        ["winget"] = "darkmagenta",
        ["chocolatey"] = "blue",
        ["scoop"] = "magenta",
    };

    public static Tree BuildDiffTree(string configFileName, NTIXConfig config, DiffResult diff)
    {
        var tree = new Tree($"[bold]{configFileName}[/]");

        if (config.Imports.Count > 0)
        {
            var importsNode = tree.AddNode("[dim]imports[/]");
            AddImportChildren(importsNode, config.Imports);
        }

        if (diff.ToInstall.Count > 0)
        {
            var node = tree.AddNode($"[green]\u2191 To install ({diff.ToInstall.Count})[/]");
            AddGroupedBySource(node, diff.ToInstall, "green", showVersion: true);
        }

        if (diff.ToUpgrade.Count > 0)
        {
            var node = tree.AddNode($"[yellow]\u2191 To upgrade ({diff.ToUpgrade.Count})[/]");
            AddGroupedBySourceWithVersions(node, diff.ToUpgrade);
        }

        if (diff.ToAdopt.Count > 0)
        {
            var node = tree.AddNode($"[cyan]\u271a To adopt ({diff.ToAdopt.Count})[/]");
            AddGroupedBySource(node, diff.ToAdopt, "cyan", showVersion: true);
        }

        if (diff.ToSkip.Count > 0)
            tree.AddNode($"[dim]\u2713 Already managed ({diff.ToSkip.Count})[/]");

        if (diff.ToRemove.Count > 0)
        {
            var node = tree.AddNode($"[red]\u2717 Orphans ({diff.ToRemove.Count})[/]");
            AddGroupedBySource(node, diff.ToRemove, "red", showVersion: false);
        }

        if (diff.IsEmpty && !diff.HasError)
            tree.AddNode("[dim]Nothing to do.[/]");

        return tree;
    }

    private static void AddImportChildren(TreeNode parent, List<ImportNode> imports)
    {
        foreach (var import in imports)
        {
            if (import.Children.Count > 0)
            {
                var node = parent.AddNode($"[dim]{import.Path}[/]");
                AddImportChildren(node, import.Children);
            }
            else
            {
                parent.AddNode($"[dim]{import.Path}[/]");
            }
        }
    }

    private static void AddGroupedBySource(TreeNode parent, List<PackageSpec> packages, string color, bool showVersion)
    {
        var grouped = packages.GroupBy(p => p.Source).OrderBy(g => g.Key);

        foreach (var group in grouped)
        {
            var sourceColor = SourceMarkup.TryGetValue(group.Key, out var c) ? c : "white";
            var count = group.Count();

            if (count == 1)
            {
                var pkg = group.First();
                var version = showVersion && pkg.Version != null ? $" ({pkg.Version})" : "";
                parent.AddNode($"[{sourceColor}]{group.Key}: {pkg.Id}{version}[/]");
            }
            else
            {
                var sourceNode = parent.AddNode($"[{sourceColor}]{group.Key} ({count})[/]");
                foreach (var pkg in group.OrderBy(p => p.Id))
                {
                    var version = showVersion && pkg.Version != null ? $" ({pkg.Version})" : "";
                    sourceNode.AddNode($"[{sourceColor}]{pkg.Id}{version}[/]");
                }
            }
        }
    }

    private static void AddGroupedBySourceWithVersions(TreeNode parent, List<PackageSpec> packages)
    {
        var grouped = packages.GroupBy(p => p.Source).OrderBy(g => g.Key);

        foreach (var group in grouped)
        {
            var sourceColor = SourceMarkup.TryGetValue(group.Key, out var c) ? c : "white";
            var count = group.Count();

            if (count == 1)
            {
                var pkg = group.First();
                parent.AddNode($"[{sourceColor}]{group.Key}: {pkg.Id} \u2192 {pkg.Version}[/]");
            }
            else
            {
                var sourceNode = parent.AddNode($"[{sourceColor}]{group.Key} ({count})[/]");
                foreach (var pkg in group.OrderBy(p => p.Id))
                {
                    sourceNode.AddNode($"[{sourceColor}]{pkg.Id} \u2192 {pkg.Version}[/]");
                }
            }
        }
    }
}

File: src/commands.rs
use colored::Colorize;
use indicatif::{ProgressBar, ProgressStyle};
use ntix_rs::config::config_loader;
use ntix_rs::execution::execution_engine;
use ntix_rs::lock::lock_file::LockFile;
use ntix_rs::models::diff_result::DiffResult;
use ntix_rs::models::options::ScoopBucket;
use ntix_rs::models::package_spec::PackageSpec;
use ntix_rs::models::{import_node::ImportNode, ntix_config::NTIXConfig};
use ntix_rs::state_management::state_service;
use ntix_rs::{diff, process_helper};
use std::path::PathBuf;
use std::time::Duration;

/// Returns `None` when a default config was just created and the caller
/// should exit early (the user still needs to edit it).
async fn resolve_and_compute(
    config_path: Option<PathBuf>,
    command_name: &str,
    adopt: bool,
    upgrade: bool,
    apply_config: bool,
) -> Result<Option<(NTIXConfig, PathBuf, DiffResult)>, Box<dyn std::error::Error>> {
    let is_new = config_path.is_none() && !config_loader::DEFAULT_CONFIG_PATH.is_file();
    let resolved_path = config_loader::ensure_default_config(config_path);

    if is_new {
        println!(
            "{}",
            format!("Created default config at {}", resolved_path.display()).green()
        );
        println!(
            "Edit it to add your packages, then run {} again.",
            format!("ntix {command_name}").bold()
        );
        return Ok(None);
    }

    let config = config_loader::load(resolved_path.clone())?;
    let state = state_service::load_state(None).unwrap_or_default();

    let config_file_name = resolved_path
        .file_name()
        .map(|n| n.to_string_lossy().into_owned())
        .unwrap_or_default();

    let spinner = ProgressBar::new_spinner();
    spinner.set_style(ProgressStyle::default_spinner().template("{spinner:.yellow} {msg}")?);
    spinner.enable_steady_tick(Duration::from_millis(100));
    spinner.set_message(config_file_name.clone().bold().to_string());

    let mut diff: DiffResult = diff::diff_engine::compute_diff(
        &config, &state, None, None, None, adopt, upgrade, true, None, &spinner,
    )
    .await?;
    spinner.finish_and_clear();

    if apply_config {
        diff::diff_engine::compute_config_files_diff(&mut diff, &config, &state);
    }

    print_diff_tree(&config_file_name, &config, &diff, apply_config);

    for w in &diff.warnings {
        eprintln!("{}", format!("Warning: {}", w).yellow());
    }

    Ok(Some((config, resolved_path, diff)))
}

pub async fn apply(
    config_path: Option<PathBuf>,
    dry_run: bool,
    no_gc: bool,
    stop_on_failure: bool,
    adopt: bool,
    upgrade: bool,
    apply_config: bool,
) -> Result<i32, Box<dyn std::error::Error>> {
    if !process_helper::is_running_as_admin() {
        eprintln!(
            "{}",
            "Error: ntix apply requires administrator privileges.".red()
        );
        eprintln!("Please re-run in an elevated terminal (Run as Administrator).");
        return Ok(1);
    }

    let Some((config, _resolved_path, mut diff)) =
        resolve_and_compute(config_path, "apply", adopt, upgrade, apply_config).await?
    else {
        return Ok(0);
    };

    if no_gc {
        diff.to_remove.clear();
    }

    if dry_run {
        println!("\n{}", "(Dry run - no changes made)".yellow());
        return Ok(0);
    }

    if diff.is_empty() {
        return Ok(0);
    }

    let mut state = state_service::load_state(None).unwrap_or_default();
    let _lock = LockFile::new(None, true)?;
    let state_path = state_service::get_state_path()?;
    let success = execution_engine::apply_diff(
        &diff,
        &config.options,
        &mut state,
        &state_path,
        stop_on_failure,
        None,
        None,
        Some(&config),
        apply_config,
        Some(&|line: &str| println!("{line}")),
        Some(&|err: &str| eprintln!("{}", err.red())),
        None,
    )
    .await;

    if success {
        println!("\n{}", "Done.".green());
        Ok(0)
    } else {
        eprintln!("\n{}", "Some operations failed.".red());
        Ok(1)
    }
}

pub async fn diff_cmd(
    config_path: Option<PathBuf>,
    adopt: bool,
    upgrade: bool,
    apply_config: bool,
) -> Result<i32, Box<dyn std::error::Error>> {
    if resolve_and_compute(config_path, "diff", adopt, upgrade, apply_config)
        .await?
        .is_none()
    {
        return Ok(0);
    }
    Ok(0)
}

pub fn state_cmd() -> Result<i32, Box<dyn std::error::Error>> {
    let Some(state) = state_service::load_state(None) else {
        println!("{}", "No state file found.".yellow());
        return Ok(0);
    };

    println!("{}", "NTIX State:".bold());

    if state.winget.is_empty() && state.chocolatey.is_empty() && state.scoop.is_empty() {
        println!("  {}", "(empty)".dimmed());
    } else {
        for (id, ver) in &state.winget {
            println!("  {}", format!("winget: {id} ({ver})").cyan());
        }
        for (id, ver) in &state.chocolatey {
            println!("  {}", format!("chocolatey: {id} ({ver})").magenta());
        }
        for (id, ver) in &state.scoop {
            println!("  {}", format!("scoop: {id} ({ver})").blue());
        }
    }

    if !state.config_files.is_empty() {
        println!("  {}", "config files:".bold());
        for (dest, hash) in &state.config_files {
            println!("    {}", format!("{dest} ({hash:.8})").white());
        }
    }

    Ok(0)
}

fn source_color(source: &str, text: &str) -> colored::ColoredString {
    match source.to_lowercase().as_str() {
        "winget" => text.truecolor(139, 0, 139),
        "chocolatey" => text.blue(),
        "scoop" => text.magenta(),
        _ => text.normal(),
    }
}

#[derive(Clone, Copy, PartialEq)]
enum Section {
    Imports,
    ToInstall,
    ToUpgrade,
    ToAdopt,
    ToSkip,
    BucketsToAdd,
    BucketsToRemove,
    ToRemove,
    ConfigFiles,
}

fn print_diff_tree(
    config_file_name: &str,
    config: &NTIXConfig,
    diff: &DiffResult,
    apply_config: bool,
) {
    let sections = [
        (Section::Imports, !config.imports.is_empty()),
        (Section::ToInstall, !diff.to_install.is_empty()),
        (Section::ToUpgrade, !diff.to_upgrade.is_empty()),
        (Section::ToAdopt, !diff.to_adopt.is_empty()),
        (Section::ToSkip, !diff.to_skip.is_empty()),
        (Section::BucketsToAdd, !diff.buckets_to_add.is_empty()),
        (Section::BucketsToRemove, !diff.buckets_to_remove.is_empty()),
        (Section::ToRemove, !diff.to_remove.is_empty()),
        (
            Section::ConfigFiles,
            apply_config
                && (!diff.config_files_to_create.is_empty()
                    || !diff.config_files_to_update.is_empty()
                    || !diff.config_files_no_longer_managed.is_empty()),
        ),
    ];
    let last_present = sections.iter().rposition(|(_, present)| *present);

    println!("{}", config_file_name.bold());

    for (i, (section, present)) in sections.iter().enumerate() {
        if *present {
            render_section(config, diff, *section, Some(i) == last_present);
        }
    }

    if diff.is_empty() {
        println!("{}", "Nothing to do.".dimmed());
    }
}

fn render_section(config: &NTIXConfig, diff: &DiffResult, section: Section, is_last: bool) {
    match section {
        Section::Imports => {
            println!(
                "{}",
                tree_branch("", is_last) + &"imports".dimmed().to_string()
            );
            print_import_children(&config.imports, &tree_continuation("", is_last));
        }
        Section::ToInstall => {
            println!(
                "{}",
                tree_branch("", is_last)
                    + &format!("\u{2191} To install ({})", diff.to_install.len())
                        .green()
                        .to_string()
            );
            print_grouped(
                &diff.to_install,
                VersionStyle::Paren,
                &tree_continuation("", is_last),
            );
        }
        Section::ToUpgrade => {
            println!(
                "{}",
                tree_branch("", is_last)
                    + &format!("\u{2191} To upgrade ({})", diff.to_upgrade.len())
                        .yellow()
                        .to_string()
            );
            print_grouped(
                &diff.to_upgrade,
                VersionStyle::Arrow,
                &tree_continuation("", is_last),
            );
        }
        Section::ToAdopt => {
            println!(
                "{}",
                tree_branch("", is_last)
                    + &format!("\u{271a} To adopt ({})", diff.to_adopt.len())
                        .cyan()
                        .to_string()
            );
            print_grouped(
                &diff.to_adopt,
                VersionStyle::Paren,
                &tree_continuation("", is_last),
            );
        }
        Section::ToSkip => {
            println!(
                "{}",
                tree_branch("", is_last)
                    + &format!("\u{2713} Already managed ({})", diff.to_skip.len())
                        .dimmed()
                        .to_string()
            );
        }
        Section::BucketsToAdd => {
            println!(
                "{}",
                tree_branch("", is_last)
                    + &format!("\u{2191} Buckets to add ({})", diff.buckets_to_add.len())
                        .green()
                        .to_string()
            );
            print_buckets(&diff.buckets_to_add, is_last);
        }
        Section::BucketsToRemove => {
            println!(
                "{}",
                tree_branch("", is_last)
                    + &format!(
                        "\u{2193} Buckets to remove ({})",
                        diff.buckets_to_remove.len()
                    )
                    .red()
                    .to_string()
            );
            print_buckets(&diff.buckets_to_remove, is_last);
        }
        Section::ToRemove => {
            println!(
                "{}",
                tree_branch("", is_last)
                    + &format!("\u{2717} Orphans ({})", diff.to_remove.len())
                        .red()
                        .to_string()
            );
            print_grouped(
                &diff.to_remove,
                VersionStyle::None,
                &tree_continuation("", is_last),
            );
        }
        Section::ConfigFiles => {
            let total = diff.config_files_to_create.len()
                + diff.config_files_to_update.len()
                + diff.config_files_no_longer_managed.len();
            println!(
                "{}",
                tree_branch("", is_last)
                    + &format!("\u{2691} Config files ({total})")
                        .white()
                        .to_string()
            );
            let prefix = tree_continuation("", is_last);
            for entry in &diff.config_files_to_create {
                println!(
                    "{}",
                    prefix.clone()
                        + "["
                        + "new".green().to_string().as_str()
                        + "] "
                        + &entry.dest.display().to_string()
                );
            }
            for entry in &diff.config_files_to_update {
                println!(
                    "{}",
                    prefix.clone()
                        + "["
                        + "update".yellow().to_string().as_str()
                        + "] "
                        + &entry.dest.display().to_string()
                );
            }
            for dest in &diff.config_files_no_longer_managed {
                println!(
                    "{}{}{}",
                    prefix.clone(),
                    "[orphan]".red(),
                    " ".to_string() + dest.as_str()
                );
            }
        }
    }
}

fn print_buckets(buckets: &[ScoopBucket], is_last: bool) {
    let mut buckets: Vec<_> = buckets.iter().collect();
    buckets.sort_by(|a, b| a.name.cmp(&b.name));
    for (j, bucket) in buckets.iter().enumerate() {
        println!(
            "{}{}",
            tree_branch(&tree_continuation("", is_last), j + 1 == buckets.len()),
            bucket.name.magenta()
        );
    }
}

fn tree_branch(prefix: &str, is_last: bool) -> String {
    format!(
        "{}{}",
        prefix,
        if is_last {
            "\u{2514}\u{2500}\u{2500} "
        } else {
            "\u{251c}\u{2500}\u{2500} "
        }
    )
}

fn tree_continuation(prefix: &str, is_last: bool) -> String {
    format!("{}{}", prefix, if is_last { "    " } else { "\u{2502}   " })
}

fn print_import_children(imports: &[ImportNode], prefix: &str) {
    let count = imports.len();
    for (i, import) in imports.iter().enumerate() {
        println!(
            "{}{}",
            tree_branch(prefix, i + 1 == count),
            import.path.display().to_string().dimmed()
        );
        if !import.children.is_empty() {
            print_import_children(&import.children, &tree_continuation(prefix, i + 1 == count));
        }
    }
}

#[derive(Clone, Copy)]
enum VersionStyle {
    None,
    Paren,
    Arrow,
}

fn print_grouped(packages: &[PackageSpec], style: VersionStyle, prefix: &str) {
    let mut sources: Vec<&str> = packages.iter().map(|p| p.source.as_str()).collect();
    sources.sort();
    sources.dedup();

    let count = sources.len();
    for (i, source) in sources.iter().enumerate() {
        let mut group: Vec<&PackageSpec> =
            packages.iter().filter(|p| p.source == *source).collect();
        group.sort_by(|a, b| a.id.cmp(&b.id));
        let source_last = i + 1 == count;
        let child_prefix = tree_continuation(prefix, source_last);

        if group.len() == 1 {
            let pkg = group[0];
            println!(
                "{}{}",
                tree_branch(prefix, source_last),
                source_color(
                    source,
                    &format!("{source}: {}{}", pkg.id, version_suffix(pkg, style))
                )
            );
        } else {
            println!(
                "{}{}",
                tree_branch(prefix, source_last),
                source_color(source, &format!("{source} ({})", group.len()))
            );
            let pkg_count = group.len();
            for (j, pkg) in group.iter().enumerate() {
                println!(
                    "{}{}",
                    tree_branch(&child_prefix, j + 1 == pkg_count),
                    source_color(source, &format!("{}{}", pkg.id, version_suffix(pkg, style)))
                );
            }
        }
    }
}

fn version_suffix(pkg: &PackageSpec, style: VersionStyle) -> String {
    let version = match pkg.version.as_deref() {
        Some(v) => v,
        None => return String::new(),
    };
    match style {
        VersionStyle::None => String::new(),
        VersionStyle::Paren => format!(" ({version})"),
        VersionStyle::Arrow => format!(" \u{2192} {version}"),
    }
}
